text/0000-unsafe-block-in-unsafe-fn.md

```diff
@@ -0,0 +1,135 @@
+calling an `unsafe` function, you must enclose that code in an `unsafe` block.
+The purpose of this is to acknowledge that the operation you are performing here
+has *not* been checked by the compiler, you are responsible yourself for
+upholding Rust's safety guarantees.  Generally, unsafe operations come with
+detailed documentation for the conditions that must be met when this operation
+is executed; it is up to you to check that all these conditions are indeed met.
+
+When you are writing a function that itself has additional conditions to ensure
+safety (say, it accesses some data without making some necessary bounds checks,
+or it takes some raw pointers as arguments and performs memory operations based
+on them), then you should mark this as an `unsafe fn` and it is up to you to
+document the conditions that must be met for the arguments.
+
+Your `unsafe fn` will likely perform unsafe operations; these have to be
+enclosed by an `unsafe` block as usual.  This is the place where you have to
+check that the requirements you documented for your own function are sufficient
+to satisfy the conditions required to perform this unsafe operation.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+First of all, we no longer warn that an `unsafe` block is unnecessary when it is
+nested immediately inside an `unsafe fn`.  So, the following compiles without
+any warning:
+
+```rust
+unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T {
+  unsafe { x.get_unchecked(i) }
+}
+```
+
+However, nested `unsafe` blocks are still redundant, so this warns:
+
+```rust
+unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T {
+  unsafe { unsafe { x.get_unchecked(i) } }
+}
+```
+
+In a next step, we have a lint that fires when an unsafe operation is performed
+inside an `unsafe fn` but outside an `unsafe` block.  So, this would trigger the
+lint:
+
+```rust
+unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T {
+  x.get_unchecked(i)
+}
+```
+
+This gets us into a state where programmers are much less likely to accidentally
+perform undesired unsafe operations inside `unsafe fn`.
+
+Even later, it might be desirable to turn this warning into an error.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This new warning will likely fire for the vast majority of `unsafe fn` out there.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+I explained the rationale in the motivation section.
+
+The alternative is to not do anything, and live with the current situation.
+
+# Prior art
+[prior-art]: #prior-art
+
+None that I am aware of: Other languages do not have `unsafe` blocks.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Should this lint be in clippy first before in becomes warn-by-default in rustc,
+to avoid a huge flood of warnings showing up at once?  Should the lint ever
+become a hard error (on newer editions), or remain a warning indefinitely?
```

Comment:
Would be nice to add a section for [Future possibilities](https://github.com/rust-lang/rfcs/pull/2561) here.
(PS: update your master branch ^.^)

In general with this RFC I would like to see more of an action plan be stated.
I appreciate that it's difficult to do it now, but as the RFC progresses I think such a plan should be encoded.
For example, a possible plan (but not necessarily the one to go with):

1. Start with:
    1. allow by default in the compiler + stop the existing compiler lint from firing.
    2. warn-by-default clippy correctness lint.
2. Make the compiler lint warn-by-default after some time.
3. In Rust 2021, we make this a hard error.

Reply:
Oh no, even more mandatory sections? I already don't get the purpose of half of the sections we have, and feel like I have to make my text less well structured to adhere to format.^^

I pretty much have an action plan in the "reference-level explanation" (which AFAIK is also about "how does this get implemented", even though that has very little to do with references for me). And my action plan matches what you said, it just doesn't have numbered steps (because those are a PITA to format in markdown, once the items are more than a single paragraph). ;)

Reply:
I added numbers to the reference section. I honestly don't know what I would put into "Future possibilities"...

Reply:
It's not *mandatory*... :) I don't think any of the sections are really mandatory... but if you can provide additional notes that is helpful. If there are future possibilities that come to mind just write that.. (but it seems like some of the unsure details re. the action plans could fit here...?)

Reply:
Nothing beyond what I wrote in "open questions" comes to my mind.